File: 01/haskell/lib/Trebuchet.hs
module Trebuchet where

import Prelude
import Data.Char

trebuchet :: String -> Int
trebuchet fileContent = (foldl to_number 0 (map (filter (not . isAlpha)) (words fileContent))) :: Int

to_number :: Int -> [Char] -> Int
to_number acc (x:[])  = acc + (read ([x] ++ [x]) :: Int)
to_number acc (x:y:[]) = acc + (read ([x] ++ [y]) :: Int)
to_number acc (head: tail) = acc + (read ([head] ++ [last tail]) :: Int)
to_number acc _ = acc + 0

File: 01/elixir/trebuchet/lib/trebuchet.ex
defmodule Trebuchet do
  @moduledoc """
  Documentation for `Trebuchet`.
  """

  @doc """
  Trebuchet

  ## Examples

      iex> Trebuchet.run(Path.join([File.cwd!,"../../fixture.txt"]))
      142

  """
  def run(file_path) do
    file_path
    |> File.read!
    |> String.split("\n", trim: true)
    |> Enum.map(&String.replace(&1, ~r/[a-z]+/, ""))
    |> Enum.map(&String.graphemes/1)
    |> Enum.map(&to_number/1)
    |> Enum.reduce(0, fn(acc, item) -> acc + item end)
  end

  def to_number(_list = [single_digit | _tail = []]) do
   String.to_integer(single_digit <> single_digit)
  end

  def to_number(_list = [head | _tail = [single_digit]]) do
    String.to_integer(head <> single_digit)
  end
    
  def to_number(_list = [head | tail]) do
    String.to_integer(head <> List.last(tail))
  end
  
end
